frontend-react/docs/proposals/permissions_layer_proposal.md

```diff
@@ -0,0 +1,216 @@
+
+---
+
+## A new approach
+
+### Addressing the solutions
+After doing some research on how permissions are handled, I have a few proposals on how to move forward.
+
+1. We are over-simplifying how we permit users by solely relying on their organization. -> We should utilize scopes and claims to better manage what permissions a user has.
+2. Relying on Okta's React library to access our authentication restricts us to only accessing our authentication information in a component. -> We should create our own Auth class that can be accessed outside of the restraints of hooks.
+
+In the code snippet of our current bearer token, the `scp` field is where we would return those scopes. We can add claims and scopes via Okta's admin UI. For example, let's say we have an `admin`, `sender`, and `receiver` scope. 
+
+| Scope       | Claims                               |
+| ----------- | ------------------------------------ |
+| admin       | data.view, data.upload, orgs.allOrgs |
+| sender      | data.upload, orgs.my-phd             |
+| receiver    | data.view, orgs.my-phd               |
+
+How these are sent through a bearer token might affect how we want to use them. For example, in Okta, we might have a `receiver` scope that serves as a baseline receiver, but perhaps a user needs additional permission to upload data. Whereas checking `if (user.isReceiver())` is our way of assuming user has the `data.read` claim, we also must give the user a `data.write` claim individually, rather than the entire `sender` scope. Then, our conditional may appear as such: `if (user.isReceiver() || user.canUpload())`
+
+#### Scope approach
+```json
+{
+  ...
+  "scp": [
+    "email",
+    "openid",
+    "receiver"
+  ]
+  ...
+}
+```
+#### Claims approach
+```json
+{
+  ...
+  "scp": [
+    "email",
+    "openid"
+  ],
+  "rs_receive": true,
+  "rs_settings": [
+      "read",
+      "write"
+  ]
+  ...
+}
+```
+
+---
+
+## Steps to achieve this
+
+### Okta configurations
+First, we would need to configure Okta with the proper claims and scopes. After that is where things get tricky! So far, ReportStream users do not have any custom information stored in their Okta user object, so we have nothing relevant to base claims off of other than what group somebody is in. 
+
+> Note:  Check with SimpleReport as they appear to have set up (or attempted to set up) similar configurations in the past.
+
+The **ONLY** thing is, while this is going on, we must keep the `organization` claim untouched as that is the basis of the entire React app's current permissions check. Once we are able to successfully provide claims to individuals, then we can implement code changes. 
+
+### The `Authentication` class
+
+This is a basic approach to how we can handle parsing claims into our own Objects. This object needs to be persisted so long as a user is logged in (likely via cookies).
```

Comment:
We don't have a server serving pages, so we don't have cookies. We can get cookies form the okta login but they aren't technically ours. We should probably save stuff in sessionStorage.

Reply:
@TomNUSDS I'm worried somebody might be able to go in and alter the object in `sessionStorage`. They could do this with cookies as well. Is there a better solution for this kind of application state persistence?

Reply:
The only way is to have the state information on the server and use a session cookie to reference it.

Encryption/signing data could work, but again that would require a server to do correctly/securely.

I wonder if Okta offers any kind of per-client data storage?